src/Service/SignalsService.php

```diff
@@ -0,0 +1,264 @@
+use App\TimeSeries\Backend\BackendException;
+use App\TimeSeries\Backend\Graphite\GraphiteBackend;
+use App\TimeSeries\Backend\Influx\InfluxBackend;
+use App\TimeSeries\TimeSeries;
+use App\Utils\QueryTime;
+
+
+class SignalsService
+{
+
+    /**
+     * @var GraphiteBackend
+     */
+    private $graphiteBackend;
+    /**
+     * @var InfluxBackend
+     */
+    private $influxBackend;
+
+    /**
+     * All available down-sample steps an datasource can use.
+     */
+    const ALLOWED_STEPS = [
+        60, 120, 300, 900, 1800, // minute-level [1, 2, 5, 15, 30]
+        3600, 7200, 21600, 43200,  //hour-level [1, 2, 6, 12]
+        86400, 172800, //day-level [1, 2]
+        604800, 1209600, 2419200, //week-level [1, 2, 4]
+        31536000, 63072000, 315360000, //year-level [1, 2, 10]
+    ];
+
+    /**
+     * Calculate step based on the value of `from` and `until`.
+     * @param QueryTime $from
+     * @param QueryTime $until
+     * @param int $maxPoints
+     * @param DatasourceEntity $datasource
+     * @return int
+     */
+    protected function calculateStep(QueryTime $from, QueryTime $until, ?int $maxPoints, DatasourceEntity $datasource){
+
+        // get the native step of the datasource
+        $nativeStep = $datasource->getNativeStep();
+
+        // if no maxPoints specified, return the native step of the datasource
+        if($maxPoints == null){
+            return $nativeStep;
+        }
+
+        // find the smallest step that can get the number of data points below the specified $maxPoints
+        $step = $nativeStep;
+        $range = $until->getEpochTime() - $from->getEpochTime();
+        foreach($this::ALLOWED_STEPS as $tmp_step){
+            if($tmp_step<$nativeStep) {
+                // ignore steps that are smaller than datasource's native step
+                continue;
+            }
+            $step = $tmp_step;
+            if($range/$tmp_step <= $maxPoints){
+                break;
+            }
+        }
+        return $step;
+    }
+
+    /**
+     * Round up the `from` time based on the step.
+     * @param $from
+     * @param $step
+     * @return QueryTime
+     */
+    private function roundUpFrom($from, $step){
+        $from_ts = $from->getEpochTime();
+        $from_ts = floor($from_ts / $step) * $step;
+        return new QueryTime($from_ts);
+    }
+
+
+    public function __construct(GraphiteBackend $graphiteBackend, InfluxBackend $influxBackend) {
+        $this->graphiteBackend = $graphiteBackend;
+        $this->influxBackend = $influxBackend;
+    }
+
+    /**
+     * Build graphite expression JSON object based on given $datasource.
+     *
+     * @param MetadataEntity $entity
+     * @param DatasourceEntity $datasource
+     * @return array
+     */
+    private function buildGraphiteExpression(MetadataEntity $entity, DatasourceEntity $datasource): array {
+        $fqid = $entity->getAttribute("fqid");
+        $queryJsons = [
+            "bgp" => [
```

Comment:
it's a pity you have to break the encapsulation and have this service also know about all possible data sources. ideally these queries belong in the datasource entity since that's the only thing that should know about the available data sources.

it would be nice to have something like `$datasource->getSignalQuery($entity)`.

however, the problem then becomes that the data source service would have to know business logic about graphite/influx in order to build the relevant part of the query.

ideally there'd be a way for the data source to not need to know the specifics of the backends, and for this service to not need to have an exhaustive list of data sources. but, i think while we are still querying two backends with such drastically different query structures this is going to be tricky to pull off.

in the current setup, the signal service needs to know about the different backends, so it's reasonable to put some business logic about them here. for graphite you'd just need the datasource to store an opaque query template which has a single placeholder (for the fqid) -- you could even have it hold the graphite query format to avoid this needless round-trip through the Charthouse Expression object. for influx you really want to know the measurement name and then a WHERE template (which might have multiple clauses i guess) -- this is a bit annoying since now you'd need to pass the entity type in order to get this information from the data source. that's not terrible, but it would mean storing repeated details for the data sources that use the graphite backend (in order not to put backend-specific logic into the data source).

@albertodainotti i think this complexity mostly goes away once you only have a single backend, so you might want to add this to the list of things to revisit once everything is in influx and you clean up this code.

Reply:
Noted added issue #4